File: shared/src/main/java/ru/privatenull/pnauth/storage/PasswordHash.java
package ru.privatenull.pnauth.storage;

public record PasswordHash(String algorithm, String salt, String hash, int iterations) {
    public PasswordHash(String salt, String hash, int iterations) {
        this("PBKDF2", salt, hash, iterations);
    }

    public static PasswordHash pbkdf2(String salt, String hash, int iterations) {
        return new PasswordHash("PBKDF2", salt, hash, iterations);
    }

    public static PasswordHash bcrypt(String hash, int cost) {
        return new PasswordHash("BCRYPT", "", hash, cost);
    }

    public static PasswordHash argon2(String hash, int iterations) {
        return new PasswordHash("ARGON2", "", hash, iterations);
    }

    public static PasswordHash legacy(String algorithm, String hash) {
        return new PasswordHash(algorithm, "", hash, 0);
    }
}

File: shared/src/main/kotlin/ru/privatenull/pnauth/storage/PasswordHash.kt
package ru.privatenull.pnauth.storage

data class PasswordHash @JvmOverloads constructor(
    val algorithm: String,
    val salt: String,
    val hash: String,
    val iterations: Int
) {
    constructor(salt: String, hash: String, iterations: Int) : this("PBKDF2", salt, hash, iterations)

    fun algorithm(): String = algorithm
    fun salt(): String = salt
    fun hash(): String = hash
    fun iterations(): Int = iterations

    companion object {
        @JvmStatic
        fun pbkdf2(salt: String, hash: String, iterations: Int): PasswordHash {
            return PasswordHash("PBKDF2", salt, hash, iterations)
        }

        @JvmStatic
        fun bcrypt(hash: String, cost: Int): PasswordHash {
            return PasswordHash("BCRYPT", "", hash, cost)
        }

        @JvmStatic
        fun argon2(hash: String, iterations: Int): PasswordHash {
            return PasswordHash("ARGON2", "", hash, iterations)
        }

        @JvmStatic
        fun legacy(algorithm: String, hash: String): PasswordHash {
            return PasswordHash(algorithm, "", hash, 0)
        }
    }
}
